Improve ArrayDeque resizing logic

isaacl opened the pull request:
Don't trim array unless it's less than 2/5 filled.
This avoids expensive resizes when an ArrayDeque hovers
around a power of 2 size.

Reimplement `trimToSize` to resize array as long as
smaller size fits data.

Fixes scala/bug#11624 crash.

isaacl commented:
cc @Ichoran @ashawley @pathikrit

ashawley commented:
Do you have experience with ScalaCheck?  Adding a property-based test for ArrayDeque would be a nice contribution to help avoid these regressions.  The JUnit test suite, as written, is clearly lacking.  I know other parts of the collections library have property tests, so you could probably use them to get started.

pathikrit commented on src/library/scala/collection/mutable/ArrayDeque.scala:
```diff
@@ -477,24 +476,33 @@ class ArrayDeque[A] protected (
 
+  @inline private[this] def shouldShrink(len: Int) = {
+    // To avoid allocation churn, only shrink when array is large
+    // and less than 2/5 filled.
```
why 2/5 not 1/2? Is this a common practice?

Ichoran replied:
You want to have your contraction threshold smaller than your expansion threshold because you want to avoid the case where you have, say, 2^m elements and add one to get 2^m + 1, causing an allocation of 2^(m+1) slots; then you remove one to go back to 2^m, and you reallocate 2^m slots.

Exactly what the threshold is varies, with no clear theory (or pragmatic data that I know of) on what is "best".  (There is a similar conundrum for fill fractions for open hashing schemes.)  One common one is 1/4, which gives a worst-case scenario of 2^(m-1) + 2^m + 2^(m+1) allocations for a cycle that removes and then adds back 2^(m-1)+1 elements (starting from 2^m+1).  But a worst case scenario for wasted space is that you consume 4x the memory that is strictly required.

In contrast, a 2/5 rule gives a worst-case reallocation of 2^m + 2^(m+1) elements after a cycle of length 2^(m-1)/5.  However, at worst you can consume 2.5x the memory that you actually need.

Is the allocation ratio of ~7x the cycle length worth the 4x memory danger?  Or is the potential 30x allocation ratio of the 2/5 scheme implausible enough so it's worth capping memory only 2.5x?  Kinda hard to say.

I don't have any particular advice, despite all the typing.

isaacl replied:
Happy to change this further if you’d suggest...

Ichoran replied:
Well you certainly need to fix the numeric overflow in `((5 * len) >>> 1)`.  `len` might be 2^30-2.  Dividing by 5 is kinda expensive, though.  (Also, you always realloc on 0; is this intentional?  It probably shouldn't be.)

isaacl replied:
I think `array.length  - 5 * (len >>> 1)  > 0` fixes the overflow in most cases?

> you always realloc on 0

I'm not sure what you mean by this

Ichoran replied:
No, that still overflows in a bad way.  If array.length is 2^30 and len is 2^30-m where m is small, 5*(len >> 1) is roughly 2^31 + 2^29, which is -2^29 signed, and 2^30 - (-2^29) doesn't overflow and is bigger than 0.  But of course you shouldn't reallocate when removing just a few elements.  The overflow _actually_ needs to be fixed, not just "in most cases"!

I also mean that the condition is always met if `len` is 0.  But reallocating a 1-length array for a 0-length array seems a bad idea, even though 0 is, to be fair, smaller than 2/5.

isaacl replied:
happy to fix the overflow issue, but to be clear I don't think my PR is introducing this bug.  The old logic checks for `2*len < array.length`.  

If `array.length` is 2^31 and len is `2^31 - m`, then `2 * len` will overflow and the check will pass.

Ichoran replied:
I thought the size of the collection was capped at 2^30 because of the bit shift mask.  You can't actually allocate an array of 2^31 (because in an `Int` 0x80000000 is -2^31), and you have to allocate powers of two.  So the old scheme avoided overflow, trickily, but probably on purpose.

isaacl committed:
Improve ArrayDeque resizing logic

Don't trim array unless it's less than 2/5 filled.
This avoids expensive resizes when an ArrayDeque hovers
around a power of 2 size.

Reimplement `trimToSize` to resize array as long as
smaller size fits data.

Fixes scala/bug#11624 crash.

isaacl commented:
please take another look!

Ichoran commented:
Fix looks good.  It's now a 1/4 threshold, but note that you could make it a 3/8 threshold by doing `(len >> 2) instead of `(len >> 1)`.  This would give a max size of 8/3 = 2 2/3 and a maximally expensive cycle of 12x, if I'm doing the math right in my head.

isaacl commented:
Threshold is equiv to `array.length > 5/2 len`, so still a 2/5.   I believe your suggestion would be equiv 4/9...

Ichoran commented:
@isaacl - Oops, you're right!  I misread the logic.
